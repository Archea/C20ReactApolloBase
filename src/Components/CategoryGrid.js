import React, { Component } from 'react'
import { graphql, Query } from 'react-apollo'
import gql from 'graphql-tag'

import { Grid, GridColumn as Column } from '@progress/kendo-react-grid'

export class CategoryGrid extends Component {
  state = {
    lastSelectedIndex: 0,
    data: []
  }
  selectionChange = event => {
    event.dataItem = {
      ...event.dataItem,
      selected: !event.dataItem.selected
    }
    this.forceUpdate()
  }

  rowClick = event => {
    let last = this.state.lastSelectedIndex
    let current = this.data.findIndex(dataItem => dataItem === event.dataItem)

    if (!event.nativeEvent.shiftKey) {
      //TODO set last selected index
      this.lastSelectedIndex = last = current
    }

    let select = !event.dataItem.selected
    for (let i = Math.min(last, current); i <= Math.max(last, current); i++) {
      this.data[i] = { ...this.data[i], selected: select }
    }
    this.forceUpdate()
  }

  headerSelectionChange = event => {
    const checked = event.syntheticEvent.target.checked
    //WHOOOOP
    this.data.forEach(item => (item.selected = checked))
    this.forceUpdate()
  }
  render() {
    if (this.props.getState && this.props.getState.loading) {
      return <div>Loading</div>
    }

    if (this.props.getState && this.props.getState.error) {
      console.log(
        'GetState for category grid error: ' + this.props.getState.error.message
      )
      return <div>Error</div>
    }

    const state = this.props
    this.data = Array.from(this.props.allProductFamilyCategories)
    console.log('you mean THIS state!?  ' + JSON.stringify(this.data))

    return (
      <div className="CategoryGrid">
        <Grid
          data={this.data}
          selectedField="selected"
          headerSelectionValue={
            this.data.findIndex(dataItem => dataItem.selected === false) === -1
          }
          onSelectionChange={this.selectionChange}
          onHeaderSelectionChange={this.headerSelectionChange}
          onRowClick={this.rowClick}>
          <Column field="name" title="Category" />
          <Column field="discipline" title="discipline" />
          <Column field="firm.name" title="firm" />
          <Column field="selected" />
        </Grid>
      </div>
    )
  }
}

const INITIALSTATE_QUERY = gql`
  query GetCategoryGridState {
    categories @client {
      lastSelectedIndex
    }
  }
`
export default graphql(INITIALSTATE_QUERY, { name: 'getState' })(CategoryGrid)
