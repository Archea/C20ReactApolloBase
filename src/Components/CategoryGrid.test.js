import React from 'react'
import ReactDOM from 'react-dom'
import { shallow, configure } from 'enzyme'
import toJSON from 'enzyme-to-json'
import { CategoryGrid } from './CategoryGrid'
import Adapter from 'enzyme-adapter-react-16'

configure({ adapter: new Adapter() })

describe('CategoryGrid', () => {
  const mockprops = {
    getState: {
      category: {
        __typename: 'Category',
        lastSelectedIndex: 0
      }
    },
    allProductFamilyCategories: [
      {
        id: 'b686f1e0-16ca-4b40-bf17-a5269f3a1cec',
        name: 'Administration',
        firm: {
          name: 'Avitru'
        },
        discipline: 'Architecture',
        selected: false
      }
    ]
  }
  test('it at least shallow renders and match snapshot', () => {
    const wrapper = shallow(<CategoryGrid {...mockprops} />)
    expect(toJSON(wrapper)).toMatchSnapshot()
  })

  const mockpropsError = {
    getState: {
      loading: false,
      error: {
        message: 'an error was mocked in your system'
      }
    }
  }
  test('it handles a query error', () => {
    const wrapper = shallow(<CategoryGrid {...mockpropsError} />)
    expect(toJSON(wrapper)).toMatchSnapshot()
  })
  const mockpropsLoading = {
    getState: {
      loading: true,
      error: null
    }
  }
  test('it handles a query loading', () => {
    const wrapper = shallow(<CategoryGrid {...mockpropsLoading} />)
    expect(toJSON(wrapper)).toMatchSnapshot()
  })
})
