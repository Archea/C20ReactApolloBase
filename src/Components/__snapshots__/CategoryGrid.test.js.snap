// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CategoryGrid it at least shallow renders and match snapshot 1`] = `
<div
  className="CategoryGrid"
>
  <KendoReactGrid
    data={
      Array [
        Object {
          "discipline": "Architecture",
          "firm": Object {
            "name": "Avitru",
          },
          "id": "b686f1e0-16ca-4b40-bf17-a5269f3a1cec",
          "name": "Administration",
          "selected": false,
        },
      ]
    }
    headerSelectionValue={false}
    onHeaderSelectionChange={[Function]}
    onRowClick={[Function]}
    onSelectionChange={[Function]}
    selectedField="selected"
  >
    <GridColumn
      field="name"
      title="Category"
    />
    <GridColumn
      field="discipline"
      title="discipline"
    />
    <GridColumn
      field="firm.name"
      title="firm"
    />
    <GridColumn
      field="selected"
    />
  </KendoReactGrid>
</div>
`;

exports[`CategoryGrid it handles a query error 1`] = `
<div>
  Error
</div>
`;

exports[`CategoryGrid it handles a query loading 1`] = `
<div>
  Loading
</div>
`;
